src/formatter.rs

```diff
@@ -242,3 +242,220 @@ impl<W: Write> Formatter for ColorFormatter<W> {
+    }
+
+    #[test]
+    fn test_color_formatter_reverse_order() {
+        // Rules don't match labels out of order
+        let mut output: Vec<u8> = vec![];
+        let mut formatter = ColorFormatter::new(
+            &mut output,
+            Arc::new(hashmap! {
+                "inner outer".to_string() => "green".to_string(),
+            }),
+        );
+        formatter.add_label("outer").unwrap();
+        formatter.add_label("inner").unwrap();
+        formatter.write_str(" hello ").unwrap();
+        formatter.remove_label().unwrap();
+        formatter.remove_label().unwrap();
+        insta::assert_snapshot!(String::from_utf8(output).unwrap(), 
+        @" hello ");
+    }
+
+    #[test]
+    fn test_color_formatter_number_of_matches_matters() {
+        // Rules that match more labels take precedence.
+        let mut output: Vec<u8> = vec![];
+        let mut formatter = ColorFormatter::new(
+            &mut output,
+            Arc::new(hashmap! {
+                "a b".to_string() => "red".to_string(),
+                "c".to_string() => "green".to_string(),
+                "b c d".to_string() => "blue".to_string(),
+            }),
+        );
+        formatter.add_label("a").unwrap();
+        formatter.write_str(" a1 ").unwrap();
+        formatter.add_label("b").unwrap();
+        formatter.write_str(" b1 ").unwrap();
+        formatter.add_label("c").unwrap();
+        formatter.write_str(" c1 ").unwrap();
+        formatter.add_label("d").unwrap();
+        formatter.write_str(" d ").unwrap();
+        formatter.remove_label().unwrap();
+        formatter.write_str(" c2 ").unwrap();
+        formatter.remove_label().unwrap();
+        formatter.write_str(" b2 ").unwrap();
+        formatter.remove_label().unwrap();
+        formatter.write_str(" a2 ").unwrap();
+        formatter.remove_label().unwrap();
+        insta::assert_snapshot!(String::from_utf8(output).unwrap(),
+        @" a1 [31m b1  c1 [34m d [31m c2  b2 [0m a2 ");
+    }
+
+    #[test]
+    fn test_color_formatter_innermost_wins() {
+        // When two labels match, the innermost one wins.
+        let mut output: Vec<u8> = vec![];
+        let mut formatter = ColorFormatter::new(
+            &mut output,
+            Arc::new(hashmap! {
+                "a".to_string() => "red".to_string(),
+                "b".to_string() => "green".to_string(),
+                "a c".to_string() => "blue".to_string(),
+                "b c".to_string() => "yellow".to_string(),
+            }),
+        );
+        formatter.add_label("a").unwrap();
+        formatter.write_str(" a1 ").unwrap();
+        formatter.add_label("b").unwrap();
+        formatter.write_str(" b1 ").unwrap();
+        formatter.add_label("c").unwrap();
+        formatter.write_str(" c ").unwrap();
+        formatter.remove_label().unwrap();
+        formatter.write_str(" b2 ").unwrap();
+        formatter.remove_label().unwrap();
+        formatter.write_str(" a2 ").unwrap();
```

Comment:
I just heard about https://lists.apache.org/thread/dz20fzls5mw8l6zlv0nmnsv6tl1qyk0t, which talks about security implications of printing unsanitized inputs (including inputs from the repo). I think we should sanitize in `ColorFormatter`. We'll just need to make sure that `jj print`[^1] prints its output to a `PlainTextFormatter` even if color was requested. Or maybe it's better to add a `write_raw/write_unescaped/write_unsafe` method to the `Formatter` trait.

[^1]: Should we rename that to `jj cat` to match `hg cat`, btw?

Reply:
I'm no expert, but it seems to me that PlainTextFormatter should also sanitize, even if that just means removing all the terminal escapes (it's probably better to replace them them like `less` without `-R` does). I think PlainTextFormatter-ed strings get printed to the terminal occasionally, and even if that's not currently the case, it would be counter-intuitive if that was more dangerous than using ColorFormatter.

So, `write_unsafe` seems to me like the way to go.

If we wanted to be serious about this, we would have to check every `println!` or `write!` inside `jj` for whether it can print untrusted data to somewhere other than a sanitizing Formatter, whether that can show up in the terminal, and if so whether that's OK for some exceptional reason (like maybe `jj print`).

Also , +1 on renaming `jj print` to `jj cat`.

Reply:
> it seems to me that PlainTextFormatter should also sanitize

Yes, that seems obvious now that you say it :)

> would have to check every `println!` or `write!`

We always write to a `Formatter` that we get from the `Ui` object, so that should already be safe. There are just a few exceptions where we print messages from library code (e.g. "ignoring git submodule foo" or "upgrading to protobuf").

Reply:
I think `jj cat`'s behavior should be switched depending on isatty(stdout). If tty, remove weird escape sequences. Otherwise dump the raw content (to e.g. a file.)

So maybe SanitizedFormatter or something, than `write_unsafe()`.

Reply:
Ah, that's a very good point. So you're saying we create a `SanitizedFormatter` that's wraps a `Formatter` and implements `Formatter` itself? Sounds good.

Reply:
or make both Sanitized/ColorFormatter go through the sanitization pass? My point is we'll need dynamic dispatch between sanitized and raw outputs.

Reply:
@yuja 's idea is clever, but the topic of security brings out a Debbie Downer in me. I feel like it's dangerous for security features to be too context-dependent. Would we really want `jj cat` to be secure while `jj cat | cat` isn't? It seems safer to have `jj cat` not be too smart and have an `--insecure` flag or something.

Reply:
I consider it as a feature (to help CLI user), not a real security issue. `cat` should dump file content unmodified, but if the output is a tty, user wouldn't care the exact content, so preprocessing is allowed.

Reply:
It makes a lot of sense -- from that perspective, there's not really a downside. It's very possible that `jj cat` isn't a big concern from the security perspective, in which case the user-friendly perspective is the only one that matters.